jax/lax/lax_control_flow.py

```diff
@@ -55,19 +57,59 @@
 _reduce = functools.reduce
 
 
-@cache()
-def _initial_style_jaxpr(fun: Callable, in_tree, in_avals):
+def _initial_style_untyped_jaxpr(fun: Callable, in_tree, in_avals):
   in_pvals = [pe.PartialVal.unknown(aval) for aval in in_avals]
   wrapped_fun, out_tree = flatten_fun_nokwargs(lu.wrap_init(fun), in_tree)
   with core.initial_style_staging():
     jaxpr, out_pvals, consts = pe.trace_to_jaxpr(
       wrapped_fun, in_pvals, instantiate=True, stage_out=False)
+  return jaxpr, out_pvals, consts, out_tree
+
+@cache()
+def _initial_style_jaxpr(fun: Callable, in_tree, in_avals):
+  jaxpr, out_pvals, consts, out_tree = _initial_style_untyped_jaxpr(
+      fun, in_tree, in_avals)
   out_avals = _map(raise_to_shaped, unzip2(out_pvals)[0])
   const_avals = tuple(raise_to_shaped(core.get_aval(c)) for c in consts)
   typed_jaxpr = core.TypedJaxpr(pe.convert_constvars_jaxpr(jaxpr),
                                 (), const_avals + in_avals, out_avals)
   return typed_jaxpr, consts, out_tree()
 
+# TODO(frostig): Cache, as in _initial_style_jaxpr, or their common subroutine
+def _initial_style_jaxprs_with_common_consts(funs: Sequence[Callable],
+                                             in_tree, in_avals):
+  jaxprs, all_out_pvals, all_consts, all_out_trees = unzip4([
+      _initial_style_untyped_jaxpr(fun, in_tree, in_avals) for fun in funs])
+
+  newvar = core.gensym('_')     # TODO(frostig): safer gensym
+  all_const_avals = tuple(
```

Comment:
Isn't there room for optimization here? If both branches close over the same value, we should take it as input only once, to reduce the wasted space for the cotangents.

Reply:
Good point. The irrelevant arguments are unused/dropped within each jaxpr, but there may still be costs. Regarding cotangents, I also wonder if there is a more broadly-applicable optimization to make in our AD, if it's not in place already?